pkgs/top-level/all-packages.nix

```diff
@@ -19467,6 +19467,49 @@ with pkgs;
 
   nixopsUnstable = lowPrio (callPackage ../tools/package-management/nixops/unstable.nix { });
 
+  /*
+   * Evaluate a NixOS configuration using this evaluation of NixPkgs.
+   *
+   * With this function you can write, for example, a package that
+   * depends on a custom virtual machine image.
+   *
+   * For the meaning of NixOS terminology like config, options and
+   * imports, see
+   * https://nixos.org/nixos/manual/index.html#sec-writing-modules
+   *
+   * Required parameters:
+   *
+   *   modules:   A list of modules that represent the configuration
+   *              of the NixOS system to be constructed. This works
+   *              like the imports attribute in NixOS modules.
+   *
+   * Optional parameters:
+   *
+   *   system:    The system architecture to use for the NixOS system.
+   *              Default: the system attribute that is already present
+   *                       in this evaluation of NixPkgs.
+   *
+   * The result is an attribute set containing:
+   *
+   *   config:    The option values resulting from evaluating the modules
+   *              modules. This contains the various outputs of the
+   *              configuration, from configuration file
+   *              details to system-spanning packages like
+   *              config.system.build.toplevel.
+   *
+   *   options:   The option declarations.
+   *
+   *   pkgs:      The NixPkgs that the system was configured to use.
+   *              This will be the NixPkgs evaluation you were using to
+   *              call this function, unless you have included a module
+   *              that overrides NixPkgs.
+   */
+  nixosEvaluate = { modules, system ? super.system }:
+    import (pkgs.path + "/nixos/lib/eval-config.nix") {
+      inherit modules system;
+      pkgs = self;
+    };
```

Comment:
Any reasons to not have short-cuts such as:
https://github.com/NixOS/nixpkgs/blob/master/nixos/default.nix#L35-L39

Reply:
`nixos/default.nix` defines three shortcuts, `system`, `vm` and `vmWithBootLoader`.

Adding `system` overloads the meaning of 'system' for (at least?) the third time. It can already mean 'system architecture' or `config.system`.

The `vm` shortcut seems arbitrary, because `vm` is only one of a number of useful attributes in `config.system.build`. `vmWithBootloader` is redundant because it only defines an option.
Also, these vm shortcuts may cause the configuration to be evaluated more than once, even when you're invoking `nixosEvaluate` only once. This can be surprising. For example, when you write a function that returns `vm` and `config`, that looks very reasonable, but `config` does not actually correspond to the configuration of `vm`.

I prefer to keep this function clean. Perhaps, maybe, something like `let result = ...eval-config...; in result // { inherit (result.config.system) build; }` might be future proof and unsurprising, but only if you agree.

Reply:
This sounds good to me, to have `{ config, options, build }`, `pkgs` should not be necessary as it is already exposed through `config._module.args.pkgs`